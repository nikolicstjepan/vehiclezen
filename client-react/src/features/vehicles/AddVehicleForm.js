import React, { useState } from "react";

import InputGroup from "../../components/InputGroup";
import ErrorAlert from "../../components/ErrorAlert";

export default function AddVehicleForm({ onAdd, error }) {
  const initialState = { make: "", model: "", year: "" };
  const [form, setForm] = useState(initialState);

  const handleFormChange = ({ target }) => {
    const { name, value } = target;

    setForm({ ...form, [name]: value });
  };

  const handleSubmit = async (e) => {
    e.preventDefault();

    try {
      await onAdd(form);
      setForm(initialState);
    } catch (error) {
      console.log({ error });
    }
  };

  return (
    <form onSubmit={handleSubmit} className="py-5 col-md-4">
      <ErrorAlert error={error} />

      <InputGroup name="make" value={form.make} onChange={handleFormChange} />
      <InputGroup name="model" value={form.model} onChange={handleFormChange} />
      <InputGroup type="number" name="year" value={form.year} onChange={handleFormChange} />

      <div className="mb-3">
        <button className="btn btn-primary">Add new</button>
      </div>
    </form>
  );
}
