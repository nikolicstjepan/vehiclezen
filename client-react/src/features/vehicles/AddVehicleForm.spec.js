import React from "react";
import renderer from "react-test-renderer";
import { shallow, mount } from "enzyme";

import AddVehicleForm from "./AddVehicleForm";

describe("AddVehicleForm", () => {
  test("snapshot renders", () => {
    const component = renderer.create(<AddVehicleForm onAdd={() => {}} />);
    let tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  test("renders inputGroup with model, make and year inputs", () => {
    const component = shallow(<AddVehicleForm onAdd={() => {}} />);

    expect(component.find("InputGroup")).toHaveLength(3);
    expect(component.find("InputGroup").at(0).prop("name")).toBe("make");
    expect(component.find("InputGroup").at(1).prop("name")).toBe("model");
    expect(component.find("InputGroup").at(2).prop("name")).toBe("year");
  });

  test("requires model, make and year to be required", () => {
    const component = mount(<AddVehicleForm onAdd={() => {}} />);

    expect(component.find("InputGroup").at(0).find("input").prop("required")).toBeTruthy();
    expect(component.find("InputGroup").at(1).find("input").prop("required")).toBeTruthy();
    expect(component.find("InputGroup").at(2).find("input").prop("required")).toBeTruthy();
  });

  test("renders button with given Add new text", () => {
    const actionText = "Add new";
    const component = shallow(<AddVehicleForm onAdd={() => {}} />);

    expect(component.find("button")).toHaveLength(1);
    expect(component.find("button").text()).toBe(actionText);
  });

  test("calls handleSubmit when button is clicked", () => {
    const onAdd = jest.fn(() => {});
    const component = mount(<AddVehicleForm onAdd={onAdd} />);

    component.find("button").simulate("submit");
    expect(onAdd).toHaveBeenCalled();
  });
});
