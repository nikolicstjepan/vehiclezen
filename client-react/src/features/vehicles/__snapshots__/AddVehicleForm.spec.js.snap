// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddVehicleForm snapshot renders 1`] = `
<form
  className="py-5 col-md-4"
  onSubmit={[Function]}
>
  <div
    className="mb-3"
  >
    <label
      className="form-label"
      htmlFor="make"
    >
      Make
    </label>
    <input
      className="form-control"
      id="make"
      name="make"
      onChange={[Function]}
      required={true}
      type="text"
      value=""
    />
  </div>
  <div
    className="mb-3"
  >
    <label
      className="form-label"
      htmlFor="model"
    >
      Model
    </label>
    <input
      className="form-control"
      id="model"
      name="model"
      onChange={[Function]}
      required={true}
      type="text"
      value=""
    />
  </div>
  <div
    className="mb-3"
  >
    <label
      className="form-label"
      htmlFor="year"
    >
      Year
    </label>
    <input
      className="form-control"
      id="year"
      name="year"
      onChange={[Function]}
      required={true}
      type="number"
      value=""
    />
  </div>
  <div
    className="mb-3"
  >
    <button
      className="btn btn-primary"
    >
      Add new
    </button>
  </div>
</form>
`;
